compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java

```diff
@@ -1608,12 +1616,26 @@ private STNode parseTypeDescriptor() {
         STNode type = parseTypeDescriptor(token.kind);
         STToken nextToken = peek();
         switch (nextToken.kind) {
-            // If next token after a type descriptor is <code>?</code> then it is an Optional type descriptor
+            //If next token after a type descriptor is <code>?</code> then it is an optional type descriptor
             case QUESTION_MARK_TOKEN:
-                return parseOptionalTypeDescriptor(type);
+                type =  parseOptionalTypeDescriptor(type);
+                break;
+            //If next token after a type descriptor is <code>[</code> then it is an array type descriptor
+            case OPEN_BRACKET_TOKEN:
+                type =  parseArrayTypeDescriptor(type);
+                break;
             default:
                 return type;
         }
+        //This part is to handle int?[] and int[]? type declarations
+        nextToken = peek();
+        if (nextToken.kind == SyntaxKind.QUESTION_MARK_TOKEN && type.kind == SyntaxKind.ARRAY_TYPE) {
+            return parseOptionalTypeDescriptor(type);
+        } else if (nextToken.kind == SyntaxKind.OPEN_BRACKET_TOKEN && type.kind == SyntaxKind.OPTIONAL_TYPE) {
+            return parseArrayTypeDescriptor(type);
+        } else {
+            return type;
+        }
```

Comment:
Can this handle  a `nilable-array or nil` scenario:  i.e: `int?[]? x;`

Reply:
Should we use recusrsion?

Reply:
Currently, it cannot handle it. Yes, recursion would be helpful, I'll fix that. Thank you.

Reply:
@SupunS I have fixed this. Please review.